### DISCUSSION TITLE
RAN550 device configuration

### DISCUSSION THREAD

**Discussion by irrenhaus** [srsran/srsRAN_Project#245] [General Help]
Hello,

since the communication with the RAN550 now seems to work on our side I'd like to ask if there is a possibility to get the configuration / configuration changes from default of the RAN550 you used with the example configs given in the repository.

Right now the RAN550 is getting the C/U traffic from srsRAN, however, srsRAN is not getting any valid packets back from the RAN550.
Since the Benetel configuration is rather confusing with many options for specific FlexRAN compatibility I'd simply like to make sure that things are configured as intended.

Specifically:
Currently the type of the RU is set to `set_ru_config_RAN550_A_2x2_Ant1_3_FlexRAN` which results in the script `/usr/sbin/radio_setup_ran550_1_a_2x2_ant1_3.sh` being executed.
However, this still sets the number of TXs to 4:
```
LOG_INFO "Set the Number of Tx's to 4 (C0300)"
/usr/bin/registercontrol -w C0300 -x 0x01010101
```

This also enables the `udCompHdr` option for PRACH (register `c0352`) while disabling PRACH compression (`C0353` is set to 0).

It does stay at a long form PRACH.

The PRACH configuration index is originally set to `0xE`, which srsRAN does not work, so we switched that to `7` as defined in the example configuration.

Then there are these two configurations:
```
LOG_INFO "Aligning TDD switching relative to downlink and uplink data and with respect to PPS (C0366)"
/usr/bin/registercontrol -w C0366 -x 0x1225
LOG_INFO "Aligning FPGA uplink timing to arrival of uplink frame(C0303)"
/usr/bin/registercontrol -w c0303 -x 0x24
```

and I'm really not sure what the correct setting for working with srsRAN would be here.

The TDD configuration `/etc/tdd.xml` was changed to
```
    <tdd-config version="1.0">
            <description>DDDDDDUUU</description>
                        <pattern index="1">
                                <slots>DDDDDDUUU</slots>
                                <numerology>1</numerology>
                                <periodicity>10</periodicity>
                                <special-slot-symbols></special-slot-symbols>
                        </pattern>
                        <pattern index="2">
                                <slots>DDDDDDUUU</slots>
                                <numerology>1</numerology>
                                <periodicity>10</periodicity>
                                <special-slot-symbols></special-slot-symbols>
                        </pattern>
    </tdd-config>
```
to match the 6-3 configuration given in the tutorial.

I'm aware that you're not the support for the RAN550 and I don't want you to give me support for the configuration, I'd just like to ask you to confirm if this is the same configuration you used when creating the example srsRAN configurations so that I can be sure that I'm matching that.

Thanks a lot for your time!

**Comment by ismagom** [comment]
In addition to this, please read amd follow RAN550 User Manual carefully. If in doubt, best would be to contact Benetel directly.

Thanks

**Comment by irrenhaus** [comment]
Hello again,

first, thanks for you help so far!

As discussed with Andre, here are the configurations we're currently using for trying to getting this to work.
For the core network we're using open5gs with its default configuration (might have changed the MCC/MNC/TAC to 001/01/1).

The behaviour currently is:
* srsRAN is connecting to the AMF without problems
* srsRAN is sending C/U plane packets to the RU and according to the `kpi` tool on the RU they're received on-time
* According to the tcpdump of these packets (
[enp177s0f0.pcap.txt](https://github.com/srsran/srsRAN_Project/files/13002588/enp177s0f0.pcap.txt), might have to rename due to file extension limitations) all IQ samples are all-0.
* The RU is not sending any packets to srsRAN (if `is_prach_cp_enabled` is set to `true` we're seeing ~2k4 packets/second on-time from the RU to srsRAN)
* We're having a look at the RF spectrum using a BladeRF SDR. According to that there is noise, a peak at the center frequency of the RU which is not related to starting/stopping traffic/srsRAN but only to the RF startup state of the RU and from time to time (not regular, roughly 5-10 seconds interval) there is a peak in activity over the whole bandwidth which does not look like there could be meaningful data in there, it looks more like there are peaks in noise coming from the RU. Example: 
![gqrx_wf_20231018_094923](https://github.com/srsran/srsRAN_Project/assets/107922/eccabaaf-333b-4e85-b7ca-9892e578f458)

* The UE cannot see the network (quite obviously I guess)

In addition to the TDD pattern that is the default on the Benetel RU and documented as matching the PRACH configuration index 159 we've also tried using configuration index 7 and the 6-3 pattern as in your example configurations - see the benetel_tdd.6-3.xml.txt file attached. There was a difference in behaviour: If that configuration index was chosen, the RU was sending ~400 packets / second to the DU MAC as soon as it booted. srsRAN was receiving and marking them as too late and warned about them being not the correct size of IQ packets and being unable to decode them correctly. It seems these packets were  due to a bug in the Benetel firmware.

If you have any ideas how we could continue to hopefully get this running it would be greatly appreciated!

Thanks
Nils / Xelera

[benetel_config.sh.txt](https://github.com/srsran/srsRAN_Project/files/13001335/benetel_config.sh.txt)
[gnb_ru_ran550_tdd_n78_100mhz_2x2.yml.txt](https://github.com/srsran/srsRAN_Project/files/13001338/gnb_ru_ran550_tdd_n78_100mhz_2x2.yml.txt)
[benetel_tdd.xml.txt](https://github.com/srsran/srsRAN_Project/files/13001340/benetel_tdd.xml.txt)
[benetel_tdd-6.3.xml.txt](https://github.com/srsran/srsRAN_Project/files/13001342/benetel_tdd-6.3.xml.txt)
[benetel_eaxc.xml.txt](https://github.com/srsran/srsRAN_Project/files/13001345/benetel_eaxc.xml.txt)

**Reply by andrepuschmann** [reply]
Hey @irrenhaus - let's try to get the DL working first and then we look at PRACH and UL. In our experiments we started with 20MHz SISO and I think you should do the same, also to keep the CPU load lower so we're not risking any missing packets due to lates. I suggest you take the default 20MHz SISO config from [here](https://github.com/srsran/srsRAN_Project/blob/main/configs/gnb_ru_ran550_tdd_n78_20mhz.yml). Note there are quite some changes, including the `iq_scaling` that could be responsible for you not seeing any DL signal in the spectrogram.

**Reply by irrenhaus** [reply]
Hello @andrepuschmann ,

thanks a lot for the help! Sadly the updated code / config leads to us having lots of realtime problems again.

Previously it was suggested to use the config snippet
```
expert:
  enable_tuned_affinity_profile: true
  number_of_threads_per_cpu: 1
  number_of_reserved_cores: 4

expert_phy:
  nof_ul_threads: 3
  nof_pdsch_threads: 5
  nof_dl_threads: 2
  max_proc_delay: 5
```

And that resolved all our realtime problems.

However, these options are gone / changed, and now I'm unable to find a configuration that does not result in me getting lots of 
`2023-11-09T14:31:26.835492 [OFH     ] [W] Late Ethernet frames detected in the transmitter queue in slot=402.14, symbol=7, type=user, direction=downlink, nof_frames=2` messages.

I'm currently trying with
```
expert_phy: 
  max_proc_delay: 5

expert_execution:
  threads: 
    ofh: 
      enable_dl_parallelization: 1
threads: 
    upper_phy: 
      pdsch_processor_type: auto
      nof_pdsch_threads: 5
# This is in the config reference but the actual application does not parse it:
#      nof_pusch_threads: 1
      nof_ul_threads: 3
      nof_dl_threads: 2
    ofh: 
      enable_dl_parallelization: 1
```

I guess it's expected that I do not see any network on the phone with that.

However: By now we have a USRP B210 in our lab and srsRAN just worked flawlessly with it @ 1x1 20MHz and our phone was able to connected to the network without issues, so we could confirm that it's at least not our UE that has a problem.

Could you give me a few hints how I can get the new version configured to not produce timing problems? The server is synchronized via PTP to a Fibrolan GM (GPS synced) and we're running a Intel Xeon Gold 6346 CPU @ 3.10GHz dual socket with 32 real cores and a XXV710 25G card for the network.

Thanks again!

**Reply by ismagom** [reply]
Hi @irrenhaus, with that Xeon CPU with 2 sockets, you should configure CPU affinities. In a dual socket, there is a high cost in communicating threads between the sockets, or migrating threads from one to another. To achieve the best performance, we need to make sure all threads run in only one socket. 

Note that, in the previous version, this might happen by coincidence. Now there is an option to force it. 

To find out how the communication cost between logical cores, you can use the command `lscpu -p`. This will show something like this:

```
# The following is the parsable format, which can be fed to other
# programs. Each different item in every column has an unique ID
# starting usually from zero.
# CPU,Core,Socket,Node,,L1d,L1i,L2,L3
0,0,0,0,,0,0,0,0
1,1,1,1,,32,32,32,1
2,2,0,0,,4,4,4,0
3,3,1,1,,36,36,36,1
4,4,0,0,,1,1,1,0
5,5,1,1,,33,33,33,1
6,6,0,0,,3,3,3,0
7,7,1,1,,35,35,35,1
8,8,0,0,,2,2,2,0
9,9,1,1,,34,34,34,1
10,10,0,0,,12,12,12,0
11,11,1,1,,44,44,44,1
12,12,0,0,,8,8,8,0
13,13,1,1,,40,40,40,1
14,14,0,0,,11,11,11,0
15,15,1,1,,43,43,43,1
16,16,0,0,,9,9,9,0
17,17,1,1,,41,41,41,1
18,18,0,0,,10,10,10,0
19,19,1,1,,42,42,42,1
20,20,0,0,,16,16,16,0
21,21,1,1,,48,48,48,1
22,22,0,0,,20,20,20,0
23,23,1,1,,52,52,52,1
24,24,0,0,,17,17,17,0
25,25,1,1,,49,49,49,1
26,26,0,0,,19,19,19,0
27,27,1,1,,51,51,51,1
28,28,0,0,,18,18,18,0
29,29,1,1,,50,50,50,1
30,30,0,0,,28,28,28,0
31,31,1,1,,60,60,60,1
32,32,0,0,,24,24,24,0
33,33,1,1,,56,56,56,1
34,34,0,0,,27,27,27,0
35,35,1,1,,59,59,59,1
36,36,0,0,,25,25,25,0
37,37,1,1,,57,57,57,1
38,38,0,0,,26,26,26,0
39,39,1,1,,58,58,58,1

```

As you can see the example above, we have 2 sockets and 2 NUMA nodes. Even cores are mapped to Socket 0 and odd cores are mapped to Socket 1. What we want here is to configure the affinity such that we either use Socket 0 or Socket 1, for instance: 

```
expert_execution:
  affinities:
    l1_dl_cpus: 16,18,20,22,24,26
    l1_ul_cpus: 28,30,32
    l2_cell_cpus: 34,36,38
    ru_cpus: 8,6,10,12,14
    low_priority_cpus: 0,2,4
  threads:
    upper_phy:
      nof_ul_threads: 2
      nof_pusch_decoder_threads: 2
      nof_pdsch_threads: 8
      nof_dl_threads: 4
    ofh:
      enable_dl_parallelization: true
``` 

uses only even cores on Socket 0.

**Reply by irrenhaus** [reply]
Hello again,

thanks a lot @ismagom, I did in fact set the affinities earlier, but it seems I chose the wrong NUMA node. Choosing the NUMA node where the NIC is connected to and upping the number up threads to the numbers you posted resolved the timing issues again.

However, this did not make any difference in terms of the RU.

I tried using a 20MHz 1x1, a 100MHz 1x1 and a 2x2 config, none of these result in the UE seeing any messages from the GNB while all parts of the system happily continue without any errors. The kpi counters in the RU still go up and say all messages are on-time.

Could you post the actual configuration of your RAN550?

The user documentation says only PRACH configuration indices 151, 156 and 159 are supported, so I'm using 151 with the 6-3 TDD pattern without a special slot.
But here the user documentation fails me: It says the prach configuration indices supported are defined by the TDD pattern being used, but what is written there does not actually work since srsRAN says that for these configs the slot indixes used for PRACH fall outside of the TDD UL slots. Also it only talks about a DDDDDDSUU configuration.

I'd like to have the setup as close as possible to what worked for you, so I'd be grateful if you could provide the tdd.xml & init script of your RU.

Thanks a lot!

**Reply by andrepuschmann** [reply]
Hey @irrenhaus - we've just done a small update to the appnote https://docs.srsran.com/projects/project/en/latest/tutorials/source/oranRU/source/index.html. Have a look at the RU section that now contains the configs we use for our device here.

**Reply by khooi8913** [reply]
Hi @irrenhaus,

Did you eventually manage to get things working with the RAN550 O-RU?
We are somewhat facing a similar issue too.
Thanks!

**Reply by ismagom** [reply]
Hi @khooi8913 , what issues are you facing exactly?

**Reply by pgawlowicz** [reply]
@irrenhaus Any update/feedback on this issue?

**Comment by khooi8913** [comment]
> Hi @khooi8913 , what issues are you facing exactly?

Hi @ismagom , I have filed my issue in a separate discussion: https://github.com/srsran/srsRAN_Project/discussions/389.
